src/expression/comparison_expression.py

```diff
@@ -32,41 +33,43 @@ def __init__(self, exp_type: ExpressionType, left: AbstractExpression,
                          children=children)
 
     def evaluate(self, *args, **kwargs):
-        # evaluate always return batch
-        left_values = self.get_child(0).evaluate(*args, **kwargs).frames
-        right_values = self.get_child(1).evaluate(*args, **kwargs).frames
+        # cast in to numpy array
+        lvalues = self.get_child(0).evaluate(*args, **kwargs).frames.values
+        rvalues = self.get_child(1).evaluate(*args, **kwargs).frames.values
 
-        if len(left_values) != len(right_values):
-            if len(left_values) == 1:
-                left_values = pd.DataFrame(np.repeat(left_values.values,
-                                                     len(right_values),
-                                                     axis=0))
-            elif len(right_values) == 1:
-                right_values = pd.DataFrame(np.repeat(right_values.values,
-                                                      len(left_values),
-                                                      axis=0))
+        if len(lvalues) != len(rvalues):
+            if len(lvalues) == 1:
+                lvalues = np.repeat(lvalues, len(rvalues), axis=0)
+            elif len(rvalues) == 1:
+                rvalues = np.repeat(rvalues, len(lvalues), axis=0)
             else:
                 raise Exception(
                     "Left and Right batch does not have equal elements")
 
         if self.etype == ExpressionType.COMPARE_EQUAL:
-            return Batch(pd.DataFrame(
-                left_values.values == right_values.values))
+            return Batch(pd.DataFrame(lvalues == rvalues))
         elif self.etype == ExpressionType.COMPARE_GREATER:
-            return Batch(pd.DataFrame(
-                left_values.values > right_values.values))
+            return Batch(pd.DataFrame(lvalues > rvalues))
         elif self.etype == ExpressionType.COMPARE_LESSER:
-            return Batch(pd.DataFrame(
-                left_values.values < right_values.values))
+            return Batch(pd.DataFrame(lvalues < rvalues))
         elif self.etype == ExpressionType.COMPARE_GEQ:
-            return Batch(pd.DataFrame(
-                left_values.values >= right_values.values))
+            return Batch(pd.DataFrame(lvalues >= rvalues))
         elif self.etype == ExpressionType.COMPARE_LEQ:
-            return Batch(pd.DataFrame(
-                left_values.values <= right_values.values))
+            return Batch(pd.DataFrame(lvalues <= rvalues))
         elif self.etype == ExpressionType.COMPARE_NEQ:
-            return Batch(pd.DataFrame(
-                left_values.values != right_values.values))
+            return Batch(pd.DataFrame(lvalues != rvalues))
+        elif self.etype == ExpressionType.COMPARE_CONTAINS:
+            res = [[all(x in p for x in q)
```

Comment:
Can we change this to work without for loops? Typically, it is slow if we use python loops. We should be able to perform this using numpy/pandas internal operations.

Reply:
Yeah, I was also trying to look for an existing sub array check in numpy. However I do not find it. 
What we can do is using set instead of array. But as far as my knowledge, I do not think we can avoid the for loop to check cell by cell.